### PR TITLE
FUSETOOLS-2475 - show a marker if the catalog version loaded does not…

### PR DESCRIPTION
**Pull request by lhein**
… match what is configured inside the pom.xml

Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

### CODE REVIEW

**Comment by lhein** [comment]
testPR

File: editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java [hunk 1]
```diff
@@ -104,4 +135,73 @@ public ProjectTemplatesSharedImages(Bundle pluginBundle) {
...
+			}
+			return false; 
+      	}
+      	
+      	private void validateConfiguredCamelVersionMatchesLoadedCamelVersion(IResource res) {
+      		Display.getDefault().asyncExec( () -> {
+		  		String loadedCatalogVersion = CamelCatalogCacheManager.getInstance().getCamelModelForProject(res.getProject()).getVersion();
+				String configuredCamelVersionFromPOM = new CamelMavenUtils().getCamelVersionFromMaven(res.getProject());
+				try {
+					listenForEvents = false;
```
Comment (apupier) [inline editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java]: if the listenForEvents attribute modification is not synchronized i think that we can be in trouble as it is an attribute which is modified in asyncExec

**Comment by apupier** [comment]
when i changed the camel version in the pom.xml to an unexisting version, i have a UI freeze for several seconds:

```
Thread [main] (Suspended)	
	owns: Object  (id=27426)	
	owns: AppInputStream  (id=27427)	
	owns: BufferedInputStream  (id=27428)	
	owns: RunnableLock  (id=27429)	
	SocketInputStream.socketRead0(FileDescriptor, byte[], int, int, int) line: not available [native method]	
	SocketInputStream.socketRead(FileDescriptor, byte[], int, int, int) line: not available	
	SocketInputStream.read(byte[], int, int, int) line: not available	
	SocketInputStream.read(byte[], int, int) line: not available	
	InputRecord.readFully(InputStream, byte[], int, int) line: not available	
	InputRecord.read(InputStream, OutputStream) line: not available	
	SSLSocketImpl.readRecord(InputRecord, boolean) line: not available	
	SSLSocketImpl.readDataRecord(InputRecord) line: not available	
	AppInputStream.read(byte[], int, int) line: not available	
	BufferedInputStream.fill() line: not available	
	BufferedInputStream.read() line: not available	
	HttpParser.readRawLine(InputStream) line: 78	
	HttpParser.readLine(InputStream, String) line: 106	
	MultiThreadedHttpConnectionManager$HttpConnectionWithReference(HttpConnection).readLine(String) line: 1116	
	MultiThreadedHttpConnectionManager$HttpConnectionAdapter.readLine(String) line: 1413	
	HeadMethod(HttpMethodBase).readStatusLine(HttpState, HttpConnection) line: 1973	
	HeadMethod(HttpMethodBase).readResponse(HttpState, HttpConnection) line: 1735	
	HeadMethod(HttpMethodBase).execute(HttpState, HttpConnection) line: 1098	
	HttpMethodDirector.executeWithRetry(HttpMethod) line: 398	
	HttpMethodDirector.executeMethod(HttpMethod) line: 171	
	HttpClient.executeMethod(HostConfiguration, HttpMethod, HttpState) line: 397	
	HttpClient.executeMethod(HttpMethod) line: 323	
	TimeoutHttpClientHandler(HttpClientHandler).doHead(URL, int) line: 266	
	TimeoutHttpClientHandler(HttpClientHandler).getURLInfo(URL, int) line: 152	
	TimeoutHttpClientHandler.getURLInfo(URL) line: 55	
	URLResource.init() line: 68	
	URLResource.exists() line: 84	
	IBiblioResolver(RepositoryResolver).findResourceUsingPattern(ModuleRevisionId, String, Artifact, ResourceMDParser, Date) line: 97	
	IBiblioResolver(AbstractPatternsBasedResolver).findResourceUsingPatterns(ModuleRevisionId, List, Artifact, ResourceMDParser, Date) line: 97	
	IBiblioResolver.findArtifactRef(Artifact, Date) line: 125	
	IBiblioResolver(BasicResolver).getArtifactRef(Artifact, Date) line: 1034	
	IBiblioResolver(BasicResolver).findFirstArtifactRef(ModuleDescriptor, DependencyDescriptor, ResolveData) line: 971	
	IBiblioResolver(BasicResolver).getDependency(DependencyDescriptor, ResolveData) line: 243	
	IBiblioResolver.getDependency(DependencyDescriptor, ResolveData) line: 506	
	ChainResolver.getDependency(DependencyDescriptor, ResolveData) line: 104	
	IvyNode.loadData(String, IvyNode, String, String, boolean, IvyNodeUsage) line: 170	
	VisitNode.loadData(String, boolean) line: 292	
	ResolveEngine.fetchDependencies(VisitNode, String, boolean) line: 718	
	ResolveEngine.doFetchDependencies(VisitNode, String) line: 803	
	ResolveEngine.fetchDependencies(VisitNode, String, boolean) line: 726	
	ResolveEngine.getDependencies(ModuleDescriptor, ResolveOptions, ResolveReport) line: 599	
	ResolveEngine.resolve(ModuleDescriptor, ResolveOptions) line: 236	
	Ivy.resolve(ModuleDescriptor, ResolveOptions) line: 523	
	Ivy$resolve$0.call(Object, Object, Object) line: not available	
	GrapeIvy.getDependencies(Map, IvyGrabRecord...) line: 410	
	GeneratedMethodAccessor153.invoke(Object, Object[]) line: not available	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	Method.invoke(Object, Object...) line: not available	
	PogoMetaMethodSite$PogoCachedMethodSite.invoke(Object, Object[]) line: 169	
	PogoMetaMethodSite$PogoCachedMethodSite(PogoMetaMethodSite).callCurrent(GroovyObject, Object[]) line: 59	
	GrapeIvy.resolve(ClassLoader, Map, List, Map...) line: 571	
	GrapeIvy$resolve$1.callCurrent(GroovyObject, Object, Object, Object, Object) line: not available	
	GrapeIvy.resolve(ClassLoader, Map, Map...) line: 538	
	GrapeIvy$resolve$0.callCurrent(GroovyObject, Object, Object, Object) line: not available	
	GrapeIvy.grab(Map, Map...) line: 256	
	GeneratedMethodAccessor138.invoke(Object, Object[]) line: not available	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	Method.invoke(Object, Object...) line: not available	
	PogoMetaMethodSite$PogoCachedMethodSite.invoke(Object, Object[]) line: 169	
	PogoMetaMethodSite$PogoCachedMethodSite(PogoMetaMethodSite).callCurrent(GroovyObject, Object[]) line: 59	
	PogoMetaMethodSite$PogoCachedMethodSite(AbstractCallSite).callCurrent(GroovyObject, Object, Object) line: 174	
	GrapeIvy.grab(Map) line: 237	
	Grape.grab(Map<String,Object>) line: 152	
	MavenVersionManager.loadVersion(String) line: 109	
	DefaultCamelCatalog.loadVersion(String) line: 163	
	CamelService.loadVersion(CamelCatalog, String) line: 113	
	CamelService.getCatalog(CamelCatalogCoordinates) line: 97	
	CamelService.updateMavenRepositoryLookup(List<Repository>, CamelCatalogCoordinates) line: 304	
	CamelManagerServiceProxy.updateMavenRepositoryLookup(List<Repository>, CamelCatalogCoordinates) line: 75	
	CamelCatalogCacheManager.getCamelModelForProject(IProject) line: 85	
	ProjectTemplatesActivator$DeltaPrinter.lambda$0(IResource) line: 180	
	1610979098.run() line: not available	
	RunnableLock.run(Display) line: 37	
	UISynchronizer(Synchronizer).runAsyncMessages(boolean) line: 182	
	Display.runAsyncMessages(boolean) line: 4213	
	Display.readAndDispatch() line: 3820	
	PartRenderingEngine$5.run() line: 1155	
	Realm.runWithDefault(Realm, Runnable) line: 336	
	PartRenderingEngine.run(MApplicationElement, IEclipseContext) line: 1044	
	E4Workbench.createAndRunUI(MApplicationElement) line: 153	
	Workbench.lambda$3(Display, WorkbenchAdvisor, int[]) line: 680	
	620301055.run() line: not available	
	Realm.runWithDefault(Realm, Runnable) line: 336	
	Workbench.createAndRunWorkbench(Display, WorkbenchAdvisor) line: 594	
	PlatformUI.createAndRunWorkbench(Display, WorkbenchAdvisor) line: 148	
	IDEApplication.start(IApplicationContext) line: 151	
	EclipseAppHandle.run(Object) line: 196	
	EclipseAppLauncher.runApplication(Object) line: 134	
	EclipseAppLauncher.start(Object) line: 104	
	EclipseStarter.run(Object) line: 388	
	EclipseStarter.run(String[], Runnable) line: 243	
	NativeMethodAccessorImpl.invoke0(Method, Object, Object[]) line: not available [native method]	
	NativeMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	DelegatingMethodAccessorImpl.invoke(Object, Object[]) line: not available	
	Method.invoke(Object, Object...) line: not available	
	Main.invokeFramework(String[], URL[]) line: 653	
	Main.basicRun(String[]) line: 590	
	Main.run(String[]) line: 1499	
	Main.main(String[]) line: 1472	
```

File: core/plugins/org.fusesource.ide.camel.model.service.impl/src/org/fusesource/ide/camel/model/service/internal/CamelService.java [hunk 2]
```diff
@@ -171,6 +171,7 @@ public CamelModel getCamelModel(String camelVersion, String runtimeProvider) {
 		CamelCatalog catalog = getCatalog(coords);
 		CamelModel loadedModel = loadCamelModelFromCatalog(catalog);
 		CamelModelPatcher.applyVersionSpecificCatalogFixes(catalog, loadedModel);
+		loadedModel.setVersion(camelVersion);
```
[the remark this answers is not in the document]
Comment (lhein) [reply core/plugins/org.fusesource.ide.camel.model.service.impl/src/org/fusesource/ide/camel/model/service/internal/CamelService.java]: The reason for doing it here is that loading the catalog changes probably the version. So what you got in catalog is probably not the camel version you asked for and hence its setting a wrong value.

File: editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java [hunk 3]
```diff
@@ -46,41 +60,57 @@ public ProjectTemplatesActivator() {
...
-    
-    @Override
-    protected BaseUISharedImages createSharedImages() {
-    	return new ProjectTemplatesSharedImages(getBundle());
-    }
-	
+
+	@Override
+	protected BaseUISharedImages createSharedImages() {
+		unregisterWorkspaceListener();
```
[the remark this answers is not in the document]
Comment (lhein) [reply editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java]: Me neither :D

File: editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java (hunk 1 above)
Comment (lhein) [reply editor/plugins/org.fusesource.ide.projecttemplates/src/org/fusesource/ide/projecttemplates/internal/ProjectTemplatesActivator.java]: good point

**Comment by lhein** [comment]
So u changed the camel version manually in the source of the pom.xml instead of using menu 'Configure > Change Camel Version' ?

**Comment by lhein** [comment]
> I have this stack written several times in log when creating a project which has no camel catalog available (such as 2.14.0):
....

I see this too. I even put in checks to only run the validation if the resource exists but for some odd reason that exception is still thrown.

**Comment by lhein** [comment]
> when i changed the camel version in the pom.xml to a version with camel catalog, i have a UI freeze for several minutes:

That seems to be caused by Ivy/Groovy downloading the catalog and its dependencies?

**Comment by lhein** [comment]
> I have this stack written several times in log when creating a project which has no camel catalog available (such as 2.14.0):

`  !ENTRY org.fusesource.ide.projecttemplates 4 0 2017-08-28 18:10:21.180
    !MESSAGE Resource '/unkownversion/bin/pom.xml' does not exist.`

fixed

**Comment by lhein** [comment]
@apupier could u give this PR another try? IMO the only thing left is your stuck UI issue comment from above. All the other issues have been addressed.

**Comment by lhein** [comment]
build failure caused by our old friend:

`org.fusesource.ide.projecttemplates.tests.integration.wizards.FuseIntegrationProjectCreatorRunnableForOSESringBootIT.testOSESpringBootProjectCreation`

Can be ignored in this PR.

**Comment by apupier** [comment]
testPR

**Comment by apupier** [comment]
> build failure caused by our old friend:
> 
> org.fusesource.ide.projecttemplates.tests.integration.wizards.FuseIntegrationProjectCreatorRunnableForOSESringBootIT.testOSESpringBootProjectCreation
> 
> Can be ignored in this PR.

better to relaunch the build to be sure that it is really not related. Especially in this case as it is touching some Camel Catalog load.

**Comment by apupier** [comment]
testPR

**Comment by apupier** [comment]
a different test has failed this time:
org.fusesource.ide.projecttemplates.tests.integration.wizards.FuseIntegrationProjectCreatorRunnableCheckCatalogCacheLoadingIT.additionalChecks(FuseIntegrationProjectCreatorRunnableCheckCatalogCacheLoadingIT.java:67)

**Comment by apupier** [comment]
testPR

**Comment by apupier** [comment]
testPR

**Comment by apupier** [comment]
it has failed again

**Comment by lhein** [comment]
#1237 is used instead

**Pull request closed without merging**